// @elderjs/elderjs/build/index.js
"use strict";
/* istanbul ignore file */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getElderConfig = exports.hookEntityDefinitions = exports.hookInterface = exports.build = exports.Elder = exports.shortcodeSchema = exports.pluginSchema = exports.routeSchema = exports.hookSchema = exports.configSchema = exports.esbuildBundler = exports.preprocessSvelteContent = exports.partialHydration = void 0;
var partialHydration_1 = require("./partialHydration/partialHydration");
Object.defineProperty(exports, "partialHydration", { enumerable: true, get: function () { return __importDefault(partialHydration_1).default; } });
Object.defineProperty(exports, "preprocessSvelteContent", { enumerable: true, get: function () { return partialHydration_1.preprocessSvelteContent; } });
var esbuildBundler_1 = require("./esbuild/esbuildBundler");
Object.defineProperty(exports, "esbuildBundler", { enumerable: true, get: function () { return __importDefault(esbuildBundler_1).default; } });
var validations_1 = require("./utils/validations");
Object.defineProperty(exports, "configSchema", { enumerable: true, get: function () { return validations_1.configSchema; } });
Object.defineProperty(exports, "hookSchema", { enumerable: true, get: function () { return validations_1.hookSchema; } });
Object.defineProperty(exports, "routeSchema", { enumerable: true, get: function () { return validations_1.routeSchema; } });
Object.defineProperty(exports, "pluginSchema", { enumerable: true, get: function () { return validations_1.pluginSchema; } });
Object.defineProperty(exports, "shortcodeSchema", { enumerable: true, get: function () { return validations_1.shortcodeSchema; } });
var Elder_1 = require("./Elder");
Object.defineProperty(exports, "Elder", { enumerable: true, get: function () { return Elder_1.Elder; } });
Object.defineProperty(exports, "build", { enumerable: true, get: function () { return Elder_1.build; } });
__exportStar(require("./utils/types"), exports);
__exportStar(require("./utils/index"), exports);
__exportStar(require("./routes/routes"), exports);
__exportStar(require("./hooks/types"), exports);
var hookInterface_1 = require("./hooks/hookInterface");
Object.defineProperty(exports, "hookInterface", { enumerable: true, get: function () { return hookInterface_1.hookInterface; } });
var hookEntityDefinitions_1 = require("./hooks/hookEntityDefinitions");
Object.defineProperty(exports, "hookEntityDefinitions", { enumerable: true, get: function () { return hookEntityDefinitions_1.hookEntityDefinitions; } });
var getConfig_1 = require("./utils/getConfig");
Object.defineProperty(exports, "getElderConfig", { enumerable: true, get: function () { return __importDefault(getConfig_1).default; } });
